glean-core/android/src/main/java/mozilla/telemetry/glean/Glean.kt

```diff
@@ -284,46 +269,66 @@ open class GleanInternalAPI internal constructor () {
+        @Suppress("EXPERIMENTAL_API_USAGE")
+        Dispatchers.API.launch {
+            val originalEnabled = internalGetUploadEnabled()
+            LibGleanFFI.INSTANCE.glean_set_upload_enabled(enabled.toByte())
+
+            if (!enabled) {
+                // Cancel any pending workers here so that we don't accidentally upload or
+                // collect data after the upload has been disabled.
+                metricsPingScheduler.cancel()
+                // Cancel any pending workers here so that we don't accidentally upload
+                // data after the upload has been disabled.
+                PingUploadWorker.cancel(applicationContext)
+            }
 
-                if (!originalEnabled && enabled) {
-                    // If uploading is being re-enabled, we have to restore the
-                    // application-lifetime metrics.
-                    initializeCoreMetrics((this@GleanInternalAPI).applicationContext)
-                }
+            if (!originalEnabled && enabled) {
+                // If uploading is being re-enabled, we have to restore the
+                // application-lifetime metrics.
+                initializeCoreMetrics((this@GleanInternalAPI).applicationContext)
+            }
 
-                if (originalEnabled && !enabled) {
-                    // If uploading is disabled, we need to send the deletion-request ping
-                    PingUploadWorker.enqueueWorker(applicationContext)
-                }
+            if (originalEnabled && !enabled) {
+                // If uploading is disabled, we need to send the deletion-request ping
+                PingUploadWorker.enqueueWorker(applicationContext)
             }
-        } else {
-            uploadEnabled = enabled
         }
     }
 
     /**
      * Get whether or not Glean is allowed to record and upload data.
+     *
+     * Caution: the result is only correct if Glean is already initialized.
+     *
+     * **THIS METHOD IS DEPRECATED.**
+     * Applications should not rely on Glean's internal state.
+     * Upload enabled status should be tracked by the application and communicated to Glean if it changes.
      */
+    @Deprecated("Upload enabled should be tracked by the application and communicated to Glean if it changes")
     fun getUploadEnabled(): Boolean {
+        return internalGetUploadEnabled()
+    }
+
+    /**
+     * Get whether or not Glean is allowed to record and upload data.
+     *
+     * Caution: the result is only correct if Glean is already initialized.
+     *
+     * Note: due to the deprecation notice and because warnings break the build,
+     * we pull out the implementation into an internal method.
+     */
+    internal fun internalGetUploadEnabled(): Boolean {
         if (isInitialized()) {
             return LibGleanFFI.INSTANCE.glean_is_upload_enabled().toBoolean()
         } else {
-            return uploadEnabled
+            return true
```

Comment:
Just looking at where this is called, this should never happen before Glean is initialized.  However, maybe it's safer to return `false` here so we err on the side of not recording things if there is a bug?  (I was also going to suggest throwing an exception -- maybe that's fine if we really consider that undefined, but having that exception accidentally "bubble out" of the API would be bad...)

Reply:
Good point.